File: apps/api/app/schemas/fairness.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel
from pydantic import ConfigDict

from app.schemas.common import TimestampedSchema
from app.schemas.schedule import SchedulePeriodRead
from app.schemas.schedule import ScheduleVersionRead


class FairnessConfigVersionRead(TimestampedSchema):
    version_number: int
    status: str
    decay_factor: float
    debt_weight: float
    favor_weight: float
    standard_priority_multiplier: float
    elevated_priority_multiplier: float
    critical_priority_multiplier: float

    model_config = ConfigDict(from_attributes=True)


class ProviderFairnessSnapshotRead(TimestampedSchema):
    provider_id: UUID
    provider_display_name: str
    schedule_period_id: UUID
    schedule_version_id: UUID
    starting_debt: float
    starting_favor_credit: float
    weekly_debt_delta: float
    weekly_favor_delta: float
    ending_debt: float
    ending_favor_credit: float
    fairness_pressure: float
    priority_tier: str
    priority_multiplier: float
    assignment_count: int
    negative_event_count: int
    positive_event_count: int


class ProviderFairnessEventRead(TimestampedSchema):
    provider_id: UUID
    provider_display_name: str
    schedule_period_id: UUID
    schedule_version_id: UUID
    assignment_id: UUID | None
    event_type: str
    category: str
    debt_delta: float
    favor_delta: float
    occurred_at: datetime
    reason: str


class FairnessMetricRead(BaseModel):
    id: str
    label: str
    value: str
    status: str
    detail: str


class FairnessReportRead(BaseModel):
    has_data: bool
    schedule_period: SchedulePeriodRead | None
    schedule_version: ScheduleVersionRead | None
    config: FairnessConfigVersionRead | None
    metrics: list[FairnessMetricRead]
    snapshots: list[ProviderFairnessSnapshotRead]
    events: list[ProviderFairnessEventRead]

File: apps/web/lib/schemas/fairness.ts
import { z } from "zod";

import {
  persistedScheduleVersionApiSchema,
  schedulePeriodApiSchema,
} from "@/lib/schemas/schedule";

export const fairnessStatusSchema = z.enum([
  "Healthy",
  "Watch",
  "Action Required",
]);

export const fairnessConfigVersionApiSchema = z.object({
  id: z.string().uuid(),
  version_number: z.number().int().min(1),
  status: z.string().min(1),
  decay_factor: z.number(),
  debt_weight: z.number(),
  favor_weight: z.number(),
  standard_priority_multiplier: z.number(),
  elevated_priority_multiplier: z.number(),
  critical_priority_multiplier: z.number(),
  created_at: z.string().min(1),
  updated_at: z.string().min(1),
});

export const providerFairnessSnapshotApiSchema = z.object({
  id: z.string().uuid(),
  provider_id: z.string().uuid(),
  provider_display_name: z.string().min(1),
  schedule_period_id: z.string().uuid(),
  schedule_version_id: z.string().uuid(),
  starting_debt: z.number(),
  starting_favor_credit: z.number(),
  weekly_debt_delta: z.number(),
  weekly_favor_delta: z.number(),
  ending_debt: z.number(),
  ending_favor_credit: z.number(),
  fairness_pressure: z.number(),
  priority_tier: z.string().min(1),
  priority_multiplier: z.number(),
  assignment_count: z.number().int().min(0),
  negative_event_count: z.number().int().min(0),
  positive_event_count: z.number().int().min(0),
  created_at: z.string().min(1),
  updated_at: z.string().min(1),
});

export const providerFairnessEventApiSchema = z.object({
  id: z.string().uuid(),
  provider_id: z.string().uuid(),
  provider_display_name: z.string().min(1),
  schedule_period_id: z.string().uuid(),
  schedule_version_id: z.string().uuid(),
  assignment_id: z.string().uuid().nullable(),
  event_type: z.string().min(1),
  category: z.string().min(1),
  debt_delta: z.number(),
  favor_delta: z.number(),
  occurred_at: z.string().min(1),
  reason: z.string().min(1),
  created_at: z.string().min(1),
  updated_at: z.string().min(1),
});

export const fairnessMetricApiSchema = z.object({
  id: z.string().min(1),
  label: z.string().min(1),
  value: z.string().min(1),
  status: fairnessStatusSchema,
  detail: z.string().min(1),
});

export const fairnessReportApiSchema = z.object({
  has_data: z.boolean(),
  schedule_period: schedulePeriodApiSchema.nullable(),
  schedule_version: persistedScheduleVersionApiSchema.nullable(),
  config: fairnessConfigVersionApiSchema.nullable(),
  metrics: z.array(fairnessMetricApiSchema),
  snapshots: z.array(providerFairnessSnapshotApiSchema),
  events: z.array(providerFairnessEventApiSchema),
});

export type FairnessStatus = z.infer<typeof fairnessStatusSchema>;

export type FairnessReportApi = z.infer<typeof fairnessReportApiSchema>;
